src/axom/core/tests/core_array_for_all.hpp

```diff
@@ -1177,4 +1177,91 @@ AXOM_TYPED_TEST(core_array_for_all, nontrivial_emplace)
   }
 }
 
+//------------------------------------------------------------------------------
+constexpr int INSERT_ON_HOST = 1;
+constexpr int INSERT_ON_DEVICE = 2;
+struct DeviceInsert
+{
+  AXOM_HOST_DEVICE DeviceInsert(int value) : m_value(value)
+  {
+#ifdef AXOM_DEVICE_CODE
+    m_host_or_device = INSERT_ON_DEVICE;
+#else
+    m_host_or_device = INSERT_ON_HOST;
+#endif
+  }
+
+  int m_value;
+  int m_host_or_device;
+};
+
+AXOM_TYPED_TEST(core_array_for_all, device_insert)
+{
+  using ExecSpace = typename TestFixture::ExecSpace;
+  using DynamicArray = typename TestFixture::template DynamicTArray<DeviceInsert>;
+  using DynamicArrayOfArrays =
+    typename TestFixture::template DynamicTArray<DynamicArray>;
+
+  int kernelAllocID = axom::execution_space<ExecSpace>::allocatorID();
+
+  constexpr axom::IndexType N = 374;
+
+  DynamicArrayOfArrays arr_container(1, 1, kernelAllocID);
+  arr_container[0] = DynamicArray(0, N, kernelAllocID);
+  const auto arr_v = arr_container.view();
+
+  EXPECT_EQ(arr_container[0].size(), 0);
+  EXPECT_EQ(arr_container[0].capacity(), N);
+
+  axom::for_all<ExecSpace>(
+    N,
+    AXOM_LAMBDA(axom::IndexType idx) {
+#ifdef AXOM_USE_OPENMP
+      if(omp_in_parallel())
+      {
+  #pragma omp critical
+        {
+          arr_v[0].emplace_back(3 * idx + 5);
+        }
+      }
+      else
+      {
+        arr_v[0].emplace_back(3 * idx + 5);
+      }
+#else
+      arr_v[0].emplace_back(3 * idx + 5);
+#endif
+    });
```

Comment:
We previously added a restriction to disallow use of `axom::Array` within a kernel  -- see https://github.com/LLNL/axom/pull/1255 and https://github.com/LLNL/axom/issues/724
It looks like this is getting around that by passing in a `axom::ArrayView` to an `axom::Array`.

* Does this feature imply that we should allow Arrays to be used in kernels?
* Alternatively, should we add `push_back` and/or `emplace_back` to an `axom::ArrayView` if/when the addition will be under capacity? Perhaps as a new `MutableArrayView` class derived from `ArrayBase` ?

Reply:
There are a few things going on here. As I see it, the main limitation with `axom::Array` use in a lambda is primarily the fact that the value-capture would copy the underlying array, instead of providing view-like semantics; for the device, there is also the issue that this "array copy" is ill-defined, since we can't allocate in device code. Thus the requirement to use `axom::ArrayView`, and the checks in #1255 being implemented in the copy constructor.

However, this doesn't necessarily preclude all uses of `axom::Array` from device code. In a certain user code, we have large hierarchies of class types that are allocated in unified memory. Some of these classes have instances of an `std::vector`-like class, and the code is able to call `push_back` on the device, which is implemented as an atomic addition on the size variable. Since the vector-like class instance is itself allocated in unified memory, this code works as intended.

I think the second proposal is interesting, but it doesn't exactly match our user code's use case; we don't wrap an externally-provided buffer with our `std::vector`-like class; rather, it owns its own memory.